File: DSA/Linked list/Rev-Rotat-Merge/RotateKtimes.py
'''
create node class
 create linl class
    create insert methode and display
     create roatesKtimes methode
    create main methode
    create linked list      
'''
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
class LinkedList:
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        last = self.head
        while last.next:
            last = last.next
        last.next = new_node

    def display(self):
        current = self.head
        while current:
            print(current.data, end=" -> ")
            current = current.next
        print("None")

    def rotate_k_times(self, k):
        if k == 0 or self.head is None:
            return

        # Count the number of nodes in the list
        count = 1
        current = self.head
        while current.next:
            count += 1
            current = current.next

        # If k is greater than the number of nodes, reduce it
        k %= count

        if k == 0:
            return

        # Find the kth node from the end
        current = self.head
        for _ in range(count - k - 1):
            current = current.next

        # Set the new head and break the list
        new_head = current.next
        current.next = None

        # Find the last node and connect it to the old head
        last_node = new_head
        while last_node.next:
            last_node = last_node.next
        last_node.next = self.head

        # Update the head to point to the new head
        self.head = new_head
if __name__ == "__main__":
    ll = LinkedList()
    ll.insert(1)
    ll.insert(2)
    ll.insert(3)
    ll.insert(4)
    ll.insert(5)

    print("Original Linked List:")
    ll.display()

    k = 2
    ll.rotate_k_times(k)

    print(f"Linked List after rotating {k} times:")
    ll.display()
    

File: DSA/Linked list/Rev-Rotat-Merge/RotateKtimes.java
/*
create node class
 create linl class
    create insert methode and display
     create roatesKtimes methode
    create main methode

 */
// class node
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

// class linked list
class LinkedList {
    Node head;

    // insert method
    public void insert(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    // rotateKtimes method
    public void rotateKtimes(int k) {
        if (head == null || k <= 0) {
            return;
        }

        Node current = head;
        int length = 1;

        // Find the length of the list
        while (current.next != null) {
            current = current.next;
            length++;
        }

        // Connect the last node to the head to make it circular
        current.next = head;
        k = k % length;
        // Find the new head after k rotations
        int stepsToNewHead = length - k;
        current = head;
        for (int i = 1; i < stepsToNewHead; i++) {
            current = current.next;
        }
        head = current.next;
        current.next = null;
    }

    // display method
    public void display() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }
}

public class RotateKtimes {
    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        // Insert elements into the linked list
        list.insert(1);
        list.insert(2);
        list.insert(3);
        list.insert(4);
        list.insert(5);

        System.out.println("Original list:");
        list.display();

        // Rotate the list k times
        int k = 2;
        list.rotateKtimes(k);

        System.out.println("List after rotating " + k + " times:");
        list.display();
    }
}